pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+)
+
+type Command struct {
+	ReplacementKeys []*NodeClaimKey
+	Candidates      []*state.StateNode
+	Reason          string    `hash:"ignore"` // reason is used for metrics
+	TimeAdded       time.Time `hash:"ignore"`
+	LastError       error     `hash:"ignore"`
+}
+
+// Hash will return a uint64 representing the hash of the replacements and candidates.
+func Hash(c *Command) (uint64, error) {
+	// Hash the command so that we can connect the two underlying data structures
+	// This only needs to be done once when it's added to the queue.
+	hash, err := hashstructure.Hash(c, hashstructure.FormatV2, &hashstructure.HashOptions{
+		SlicesAsSets:    true,
+		IgnoreZeroValue: true,
+		ZeroNil:         true,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("hashing command, %w", err)
+	}
+	return hash, nil
+}
+
+type NodeClaimKey struct {
+	nodeclaim.Key
+	Initialized bool // wrap a bool so we can save on recurring calls to check for initialization
+}
+
+type Queue struct {
+	workqueue.RateLimitingInterface
+	candidateProviderIDToCommandID map[string]uint64   // providerID -> commandID, used for quick checks to see if an individual candidate is in queue
+	idToCommand                    map[uint64]*Command // commandID -> Command, used to associate a unique identifier to the Command details.
+
+	kubeClient client.Client
+	recorder   events.Recorder
+	cluster    *state.Cluster
+	clock      clock.Clock
+}
+
+// NewQueue creates a queue that will asynchronously orchestrate deprovisioning commands
+func NewQueue(ctx context.Context, kubeClient client.Client, recorder events.Recorder, cluster *state.Cluster, clock clock.Clock) *Queue {
+	queue := &Queue{
+		RateLimitingInterface:          workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(queueBaseDelay, queueMaxDelay)),
+		candidateProviderIDToCommandID: map[string]uint64{},
+		idToCommand:                    map[uint64]*Command{},
+		kubeClient:                     kubeClient,
+		recorder:                       recorder,
+		cluster:                        cluster,
+		clock:                          clock,
+	}
+	go queue.Start(logging.WithLogger(ctx, logging.FromContext(ctx).Named("deprovisioning")))
+	return queue
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, reason string, timeAdded time.Time) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates: candidates,
+		Reason:     reason,
+		TimeAdded:  timeAdded,
+	}
+}
+
+// Start will kick off the queue. This runs asynchronously unil Karpenter shuts down.
+func (q *Queue) Start(ctx context.Context) {
+	for {
+		// Get command from queue. This waits until queue is non-empty.
+		item, shutdown := q.RateLimitingInterface.Get()
+		if shutdown {
+			break
+		}
+		hash := item.(uint64)
```

Comment:
I'm not quite getting why we need to hash the commands like this. Can't we just use the pointer to the command as an identifier and then track the providerIDs that are part of any given active command separately?